pennylane/decomposition/controlled_decomposition.py

```diff
@@ -0,0 +1,239 @@
+            base_params=base_op_rep.params,
+            num_control_wires=num_control_wires,
+            num_zero_control_values=0,
+            num_work_wires=num_work_wires,
+        )
+        if rep.params["num_control_wires"] == 1:
+            return resource_rep(qml.CNOT)
+        if rep.params["num_control_wires"] == 2:
+            return resource_rep(qml.Toffoli)
+        return resource_rep(
+            qml.MultiControlledX,
+            num_control_wires=rep.params["num_control_wires"],
+            num_zero_control_values=rep.params["num_zero_control_values"],
+            num_work_wires=rep.params["num_work_wires"],
+        )
+
+    return controlled_resource_rep(
+        base_class=base_op_rep.op_type,
+        base_params=base_op_rep.params,
+        num_control_wires=num_control_wires,
+        num_zero_control_values=0,
+        num_work_wires=num_work_wires,
+    )
+
+
+class GeneralControlledDecomposition(DecompositionRule):  # pylint: disable=too-few-public-methods
+    """A decomposition rule for a controlled operation with a decomposition."""
+
+    def __init__(self, base_decomposition: DecompositionRule):
+        self._base_decomposition = base_decomposition
+        super().__init__(self._get_impl(), self._get_resource_fn())
+
+    def _get_impl(self) -> Callable:
+        """The default implementation of a controlled decomposition."""
+
+        def _impl(*_, control_wires, control_values, work_wires, base, **__):
+            for w, val in zip(control_wires, control_values):
+                if not val:
+                    qml.PauliX(w)
+            qml.ctrl(
+                self._base_decomposition._impl,  # pylint: disable=protected-access
+                control=control_wires,
+                work_wires=work_wires,
+            )(*base.parameters, wires=base.wires, **base.hyperparameters)
+            for w, val in zip(control_wires, control_values):
+                if not val:
+                    qml.PauliX(w)
+
+        return _impl
+
+    def _get_resource_fn(self) -> Callable:
+        """The resource function."""
+
+        def _resource_fn(
+            *_, base_params, num_control_wires, num_zero_control_values, num_work_wires, **__
+        ):
+            base_resource_decomp = self._base_decomposition.compute_resources(**base_params)
+            gate_counts = {
+                _controlled_resource_rep(base_op_rep, num_control_wires, num_work_wires): count
+                for base_op_rep, count in base_resource_decomp.gate_counts.items()
+            }
+            gate_counts[resource_rep(qml.PauliX)] = num_zero_control_values * 2
+            return gate_counts
+
+        return _resource_fn
+
+
+def _controlled_g_phase_resource(
+    *_, num_control_wires, num_zero_control_values, num_work_wires, **__
+):
+    if num_control_wires == 1:
+        return {
+            resource_rep(qml.PauliX): num_zero_control_values * 2,
+            resource_rep(qml.PhaseShift): 1,
```

Comment:
I think we should be decomposing this as a `GlobalPhase` and `PhaseShift` if the control value is 0, not as 2 `PauliX` and a `PhaseShift`. That's a bit off-topic maybe :thinking:

Reply:
```pycon
>>> phi = 0.3
>>> mat_orig = qml.ctrl(qml.GlobalPhase(phi, wires=[1]), control=[0], control_values=[0]).matrix()
>>> mat_decomp = qml.matrix(qml.GlobalPhase(phi, wires=[0,1]) @ qml.PhaseShift(phi, 0), wire_order=[0,1])
>>> np.allclose(mat_orig, mat_decomp)
True
```

Reply:
Nice, will update.